wmfdata/spark.py

```diff
@@ -112,20 +131,41 @@ def get_custom_session(
     """
     check_kerberos_auth()
 
+    if SPARK_HOME == "/usr/lib/spark2":
+        warnings.warn(
+            "\n    Spark 2 has been deprecated. Please migrate to Spark 3."
+            "\n    See https://wikitech.wikimedia.org/wiki/Analytics/Systems/Cluster/Spark/"
+            "Migration_to_Spark_3",
+            category=FutureWarning
+        )
+
+    old_session = get_active_session()
+    if old_session:
+        old_session.stop()
+
     if master == "yarn":
         if ship_python_env:
             # The path to our packed conda environment.
             conda_packed_file = conda.pack(**conda_pack_kwargs)
             # This will be used as the unpacked directory name in the YARN working directory.
             conda_packed_name = os.path.splitext(os.path.basename(conda_packed_file))[0]
 
-            # Ship conda_packed_file to each YARN worker.
+            # Ship conda_packed_file to each YARN worker, unless a previous session has done it already.
             conda_spark_archive = f"{conda_packed_file}#{conda_packed_name}"
-            if "spark.yarn.dist.archives" in spark_config:
-                spark_config["spark.yarn.dist.archives"] += f",{conda_spark_archive}"
+            # Hack: For some reason, the SparkConf associated with a previous SparkContext persists.
+            # Therefore, we should only set 'spark.yarn.dist.archives' if the SparkConf doesn't have it
+            # as otherwise we will get a dist cache exception.
+            # see https://github.com/wikimedia/wmfdata-python/pull/36#discussion_r1023307058
+            is_archive_set = old_session and \
+                old_session.sparkContext.getConf().contains("spark.yarn.dist.archives")
+            if not is_archive_set:
+                if "spark.yarn.dist.archives" in spark_config:
+                    spark_config["spark.yarn.dist.archives"] += f",{conda_spark_archive}"
+                else:
+                    spark_config["spark.yarn.dist.archives"] = conda_spark_archive
```

Comment:
Hmm, you still get the "attempt to add X multiple times to the distributed cache" error if you:
1. Create a session with a shipped environment
2. Manually stop the session
3. Attempt to create a new session with a shipped environment

There's no `old_session` since it was already stopped, so it attempts to ship the packed environment again.

Test case and error: https://phabricator.wikimedia.org/P40122

Reply:
Ahh, I think I see a way to fix this. Commit incoming!

Reply:
Isn't this a far fetched failure case? Either we let the new APIs take care of the `SparkSession` lifecycle, or we don't. And if we don't, then we don't need these new APIs. Right?

We could make this explicit on the API docs.

Reply:
It won't be that common, but I don't think it's far-fetched. We definitely intend for people to create sessions and then use the session methods directly, and one of those methods is `stop()`. Besides, sessions sometimes die unexpectedly due to errors, and that would leave us in the same state (archive previously shipped but no active session).

Reply:
Fair enough.

I had discussed with @ottomata the possibility of just creating a toss away `SparkConf`. That would be independent of the session. So a potential solution for the current issue.

Reply:
> I had discussed with @ottomata the possibility of just creating a toss away SparkConf. That would be independent of the session. So a potential solution for the current issue.

Yeah, I was thinking about that too, but I don't actually see a way to do that. If you call `pyspark.SparkConf()` you still get the old settings rather than  a fresh config. But I think [be853b0](https://github.com/wikimedia/wmfdata-python/pull/36/commits/be853b0b59cc74f30c1ee464a34b11663b1c9afe) solves it—take a look and tell me what you think!

Reply:
Perfect!